File: neet_code_150_cpp/math_geometry/happy_number.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <unordered_set>

//AC
class Solution {
    public:
        bool is_happy(int n) {
            auto m = n;
            std::unordered_set<int> set;

            while(m != 1) {
                std::vector<int> digits;
                while(m > 0) {
                    digits.push_back(m % 10);
                    m /= 10;
                }
                m = sq_sum(digits);
                if(set.find(m) != set.end()) {
                    return false;
                } else {
                    set.insert(m);
                }
            }

            return true;
        }

    private:
        int sq_sum(const std::vector<int>& vec) {
            int sq_sum = 0;
            for(const auto& x: vec) {
                sq_sum += x * x;
            }
            return sq_sum;
        }
};

// 模範解答
class SolutionAns {
    public:
        bool isHappy(int n) {
            int slow = n;
            int fast = getNext(n);

            while (slow != fast && fast != 1) {
                slow = getNext(slow);
                fast = getNext(getNext(fast));
            }

            if (fast == 1) {
                return true;
            }
            return false;
        }
    private:
        int getNext(int n) {
            int sum = 0;
            while (n > 0) {
                int digit = n % 10;
                n /= 10;
                sum += std::pow(digit, 2);
            }
            return sum;
        }
};

int main(void) {
    Solution s_1;
    std::cout << "case_1: " << s_1.is_happy(19) << std::endl;
    std::cout << "case_2: " << s_1.is_happy(2) << std::endl;
    std::cout << "case_3: " << s_1.is_happy(12) << std::endl;

    SolutionAns s_2;
    std::cout << "case_1: " << s_2.isHappy(19) << std::endl;
    std::cout << "case_2: " << s_2.isHappy(2) << std::endl;
    std::cout << "case_3: " << s_2.isHappy(12) << std::endl;
}

File: neet_code_150/src/math_and_geometry/happy_number.rs
use std::collections::HashSet;

// AC
// しかし、速度、メモリともに悪い
struct Solution {}
impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut m = n;
        let mut set = HashSet::new();

        while m != 1 {
            let mut digits = vec![];
            while m > 0 {
                digits.push(m % 10);
                m /= 10;
            }
            m = digits.iter().map(|x| x * x).sum();
            // mが同じ数が複数回現れると無限ループとなってしまう
            if set.contains(&m) {
                return false;
            } else {
                set.insert(m);
            }
            /*
            println!("digits: {:?}", digits);
            println!("m: {}", m);
            */
        }

        true
    }
}

// 模範解答
// match式の "1|4"が??
struct SolutionAns {}
impl SolutionAns {
    pub fn is_happy(mut n: i32) -> bool {
        loop {
            let mut s = 0;
            while n > 0 {
                s += (n % 10).pow(2);
                n /= 10;
            }
            match s {
                1 | 4 => break s == 1,
                _ => n = s,
            }
        }
    }
}

fn main() {
    println!("case_1: {}", Solution::is_happy(19));
    println!("case_2: {}", Solution::is_happy(2));
    println!("case_3: {}", Solution::is_happy(12));

    println!("case_1: {}", SolutionAns::is_happy(19));
    println!("case_2: {}", SolutionAns::is_happy(2));
    println!("case_3: {}", SolutionAns::is_happy(12));
}
